File: platforms/android/unitrack/src/main/java/com/unitrack/sdk/DeviceInfo.kt
package com.unitrack.sdk

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.provider.Settings
import org.json.JSONObject
import java.io.File
import java.util.Locale
import java.util.TimeZone

/**
 * Collects device + app metadata attached to every event (parity with what
 * Firebase auto-collects). Gathered once at init and handed to the C core via
 * NativeBridge.setDeviceInfo.
 *
 * `network_type` requires the ACCESS_NETWORK_STATE permission (declared in the
 * library manifest). If unavailable it degrades to "unknown" — never throws.
 */
internal object DeviceInfo {

    fun json(app: Application): String {
        val o = JSONObject()
        runCatching {
            val pkg = app.packageName
            val pm = app.packageManager
            val pi = pm.getPackageInfo(pkg, 0)
            // User-facing app title (android:label on the application tag in
            // AndroidManifest). Same idea as iOS CFBundleDisplayName — show
            // "Mobi X Staging" not just the package name.
            val appName = runCatching { pm.getApplicationLabel(app.applicationInfo)?.toString() ?: "" }
                .getOrDefault("")

            // Field naming aligned with FPT application_context Iglu schema —
            // cross-platform keys match iOS so 1 schema validator + 1 SQL
            // query work for both platforms. Legacy aliases kept for
            // in-flight portal consumers.
            //   bundle           = packageName (iOS bundle_id sibling)
            //   device_model     = "Android" (generic family, same as iOS "iPhone"/"iPad")
            //   device_name      = "Samsung Galaxy S25 Ultra" (marketing-style label)
            //   device_imei      = ANDROID_ID (we don't request READ_PHONE_STATE)
            //   versioncode      = PackageInfo.versionCode (iOS CFBundleVersion sibling)
            //   network_label    = "wifi" / "5G" / "4G" / "3G" / "2G" / "ethernet"
            //   network_strength = WifiManager.connectionInfo.rssi (Wi-Fi only)
            val deviceId = androidId(app)
            val net = networkSnapshot(app)
            o.put("platform", "android")
            o.put("os", "Android")
            o.put("os_version", Build.VERSION.RELEASE ?: "")
            o.put("api_level", Build.VERSION.SDK_INT)
            // device_model = generic family; device_name = marketing label
            o.put("device_model", "Android")
            o.put("device_name", marketingDeviceName())
            o.put("model", Build.MODEL ?: "")                 // alias - hardware id
            o.put("manufacturer", Build.MANUFACTURER ?: "")
            o.put("app_name", appName)
            o.put("app_version", pi.versionName ?: "")
            o.put("app_build", versionCode(pi).toString())    // alias
            o.put("versioncode", versionCode(pi).toString())  // schema key
            // Cross-platform bundle key + legacy aliases
            o.put("bundle", pkg)                              // schema key
            o.put("app_bundle", pkg)                          // legacy alias
            o.put("app_package", pkg)                         // legacy alias
            o.put("bundle_id", pkg)                           // legacy alias
            o.put("locale", Locale.getDefault().toString())
            o.put("timezone", TimeZone.getDefault().id)
            o.put("screen", "${app.resources.displayMetrics.widthPixels}x" +
                            "${app.resources.displayMetrics.heightPixels}@" +
                            "${app.resources.displayMetrics.density}x")
            // Network — transport + friendly label + signal strength.
            o.put("network_type", net["type"] ?: "unknown")
            o.put("cellular_subtype", net["cellular_subtype"] ?: "")
            o.put("network_label", net["label"] ?: "")
            o.put("network_strength", net["strength"] ?: "")
            // Iglu schema application_context/1-0-0 khai 2 field này là string
            // ("true"/"false"), không phải boolean. Nếu put Boolean thẳng vào
            // JSONObject → validator Snowplow reject vào bad-events với
            // ValidationError: "boolean found, string expected". iOS đã cast
            // sẵn, giữ parity ở đây.
            o.put("is_debug", isDebuggable(app).toString())
            o.put("is_rooted", isRooted().toString())
            // device_imei: Android cấm READ_PHONE_STATE từ API 29+ trừ system app.
            // SDK fill bằng ANDROID_ID (Settings.Secure.ANDROID_ID) — stable
            // per app signing key + device, reset khi factory reset.
            o.put("device_imei", deviceId)                    // schema key
            o.put("device_id", deviceId)                      // legacy alias
            o.put("sdk_version", "1.0.0")
        }
        return o.toString()
    }

    @Suppress("DEPRECATION")
    private fun versionCode(pi: android.content.pm.PackageInfo): Long =
        if (Build.VERSION.SDK_INT >= 28) pi.longVersionCode else pi.versionCode.toLong()

    private fun isDebuggable(app: Application): Boolean =
        (app.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) != 0

    private fun androidId(app: Application): String =
        runCatching {
            Settings.Secure.getString(app.contentResolver, Settings.Secure.ANDROID_ID) ?: ""
        }.getOrDefault("")

    /** Marketing name from Build.MANUFACTURER + MODEL. Examples:
     *    "Samsung Galaxy S25 Ultra"
     *    "Google Pixel 9 Pro"
     *    "Xiaomi 14 Pro"
     *  Some OEMs already include the manufacturer in MODEL — we dedupe to
     *  avoid "Samsung Samsung Galaxy …".
     */
    private fun marketingDeviceName(): String {
        val manu = (Build.MANUFACTURER ?: "").trim()
        val model = (Build.MODEL ?: "").trim()
        if (manu.isEmpty()) return model
        if (model.isEmpty()) return manu
        if (model.lowercase(Locale.ROOT).startsWith(manu.lowercase(Locale.ROOT))) return model
        // Capitalize manufacturer first letter so "samsung Galaxy" → "Samsung Galaxy"
        val capManu = manu.replaceFirstChar { it.uppercase(Locale.ROOT) }
        return "$capManu $model"
    }

    /** Snapshot the active network. Returns a map with:
     *    type             — wifi | cellular | ethernet | none | unknown
     *    cellular_subtype — 2g | 3g | 4g | 5g (only when type=cellular)
     *    label            — friendly tag: "wifi" / "5G" / "4G" / "ethernet"
     *    strength         — Wi-Fi RSSI dBm (vd "-55"); empty for cellular
     *                       because Android doesn't expose RSSI without
     *                       READ_PHONE_STATE permission for non-system apps.
     */
    private fun networkSnapshot(app: Application): Map<String, String> = runCatching {
        val cm = app.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return@runCatching mapOf("type" to "unknown")
        val net = cm.activeNetwork ?: return@runCatching mapOf("type" to "none")
        val caps = cm.getNetworkCapabilities(net) ?: return@runCatching mapOf("type" to "none")
        val out = mutableMapOf<String, String>()
        when {
            caps.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> {
                out["type"] = "wifi"
                out["label"] = "wifi"
                wifiRssi(app)?.let { out["strength"] = it.toString() }
            }
            caps.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> {
                val gen = cellularGeneration(app)
                out["type"] = "cellular"
                out["cellular_subtype"] = gen
                out["label"] = gen.uppercase()
            }
            caps.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> {
                out["type"] = "ethernet"
                out["label"] = "ethernet"
            }
            else -> {
                out["type"] = "other"
                out["label"] = "other"
            }
        }
        out
    }.getOrDefault(mapOf("type" to "unknown"))

    /** Wi-Fi RSSI in dBm via WifiManager. Returns null if location permission
     *  is missing (Android 10+ requires it for connectionInfo). */
    private fun wifiRssi(app: Application): Int? = runCatching {
        val wm = app.applicationContext
            .getSystemService(Context.WIFI_SERVICE) as? android.net.wifi.WifiManager ?: return null
        @Suppress("DEPRECATION")
        wm.connectionInfo?.rssi?.takeIf { it != -127 }  // -127 = no connection / no permission
    }.getOrNull()

    /** Map the cellular data network type to a generation (2g/3g/4g/5g). */
    @Suppress("DEPRECATION")
    private fun cellularGeneration(app: Application): String = runCatching {
        val tm = app.getSystemService(Context.TELEPHONY_SERVICE)
            as? android.telephony.TelephonyManager ?: return@runCatching "cellular"
        // getDataNetworkType (API 30+) needs no permission; networkType is the
        // legacy fallback. A SecurityException degrades to plain "cellular".
        val type = try {
            if (Build.VERSION.SDK_INT >= 30) tm.dataNetworkType else tm.networkType
        } catch (_: SecurityException) {
            return@runCatching "cellular"
        }
        when (type) {
            android.telephony.TelephonyManager.NETWORK_TYPE_GPRS,
            android.telephony.TelephonyManager.NETWORK_TYPE_EDGE,
            android.telephony.TelephonyManager.NETWORK_TYPE_CDMA,
            android.telephony.TelephonyManager.NETWORK_TYPE_1xRTT,
            android.telephony.TelephonyManager.NETWORK_TYPE_IDEN -> "2g"
            android.telephony.TelephonyManager.NETWORK_TYPE_UMTS,
            android.telephony.TelephonyManager.NETWORK_TYPE_EVDO_0,
            android.telephony.TelephonyManager.NETWORK_TYPE_EVDO_A,
            android.telephony.TelephonyManager.NETWORK_TYPE_EVDO_B,
            android.telephony.TelephonyManager.NETWORK_TYPE_HSDPA,
            android.telephony.TelephonyManager.NETWORK_TYPE_HSUPA,
            android.telephony.TelephonyManager.NETWORK_TYPE_HSPA,
            android.telephony.TelephonyManager.NETWORK_TYPE_HSPAP,
            android.telephony.TelephonyManager.NETWORK_TYPE_EHRPD -> "3g"
            android.telephony.TelephonyManager.NETWORK_TYPE_LTE -> "4g"
            android.telephony.TelephonyManager.NETWORK_TYPE_NR -> "5g"
            else -> "cellular"
        }
    }.getOrDefault("cellular")

    /** Best-effort root detection: su binaries / known root packages / test-keys. */
    private fun isRooted(): Boolean {
        if (Build.TAGS?.contains("test-keys") == true) return true
        val paths = arrayOf(
            "/system/app/Superuser.apk", "/sbin/su", "/system/bin/su",
            "/system/xbin/su", "/data/local/xbin/su", "/data/local/bin/su",
            "/system/sd/xbin/su", "/system/bin/failsafe/su", "/data/local/su",
            "/su/bin/su", "/magisk"
        )
        return paths.any { runCatching { File(it).exists() }.getOrDefault(false) }
    }
}

File: platforms/react-native/ios/Native/swift/DeviceInfo.swift
// DeviceInfo.swift
//
// Collects device + app metadata attached to every event (parity with what
// Firebase auto-collects). Gathered once at init and handed to the C core via
// ut_set_device_info; the core then stamps it onto each event payload.

import Foundation
import UIKit
import CoreTelephony
import Network

enum DeviceInfo {

    /// A JSON object string: {"os":"iOS","os_version":"17.4", ...}
    static func json() -> String {
        let dev = UIDevice.current
        let bundle = Bundle.main
        let info = bundle.infoDictionary ?? [:]
        let screen = UIScreen.main.bounds.size
        let scale = UIScreen.main.scale

        // User-facing app title: prefer CFBundleDisplayName (e.g. "Mobi X
        // Staging"), fall back to CFBundleName, then bundle name. So the
        // operator sees the same label they'd see under the home-screen icon
        // instead of just the bundle id.
        let appName = str(info["CFBundleDisplayName"]).isEmpty
            ? (str(info["CFBundleName"]).isEmpty ? "" : str(info["CFBundleName"]))
            : str(info["CFBundleDisplayName"])
        let bundleId = bundle.bundleIdentifier ?? ""
        let net = networkSnapshot()

        // Field naming aligned with the FPT application_context Iglu schema —
        // cross-platform fields use the same key on iOS + Android so the
        // schema validator + downstream queries don't need to fork.
        //   bundle        — iOS bundle id, Android packageName
        //   device_model  — generic family ("iPhone"/"iPad" on iOS, "Android")
        //   device_name   — marketing label ("iPhone 15 Pro", "Samsung Galaxy S25 Ultra")
        //   device_imei   — IDFV on iOS (Apple won't give the real IMEI),
        //                   ANDROID_ID on Android. Stable per install.
        //   versioncode   — CFBundleVersion on iOS, PackageInfo.versionCode on Android
        //   network_type  — wifi | cellular | wired | vpn | none
        //   network_label — combined human label "wifi" / "5G" / "4G" / "3G" / "2G"
        //   network_strength — RSSI dBm (Wi-Fi) or signal bars (cellular)
        let deviceId = identifierForVendor()
        let fields: [String: String] = [
            "platform":       "ios",
            "os":             dev.systemName,                                  // "iOS"
            "os_version":     dev.systemVersion,                               // "17.4"
            // device_model = generic family ("iPhone"/"iPad"); device_name =
            // marketing label ("iPhone 15 Pro"). Older `model` key kept as
            // alias for in-flight downstream consumers.
            "device_model":   dev.model,                                      // "iPhone"
            "device_name":    marketingDeviceName(),                          // "iPhone 15 Pro"
            "model":          hardwareModel(),                                // "iPhone15,2" (alias)
            "manufacturer":   "Apple",
            "app_name":       appName,                                        // "Mobi X Staging"
            "app_version":    str(info["CFBundleShortVersionString"]),        // "1.0.0"
            "app_build":      str(info["CFBundleVersion"]),                   // "42" (alias)
            "versioncode":    str(info["CFBundleVersion"]),                   // unified cross-platform key
            // Cross-platform bundle key. `app_bundle` + `bundle_id` are
            // legacy aliases — same value, kept until every consumer migrates.
            "bundle":         bundleId,                                       // schema key
            "app_bundle":     bundleId,                                       // legacy alias
            "bundle_id":      bundleId,                                       // legacy alias
            "locale":         Locale.current.identifier,                       // "vi_VN"
            "timezone":       TimeZone.current.identifier,
            "screen":         "\(Int(screen.width))x\(Int(screen.height))@\(Int(scale))x",
            // Network: NWPathMonitor + CoreTelephony + WiFi RSSI estimate.
            //   network_type     — wifi | cellular | wired | vpn | none (transport)
            //   cellular_subtype — 2g | 3g | 4g | 5g (filled when cellular)
            //   network_label    — "wifi" / "5G" / "4G" / ... (1 friendly label)
            //   network_strength — RSSI (Wi-Fi, dBm, vd "-55"); signal bars (cell, "0".."4")
            //                      Empty when the value isn't accessible (no
            //                      Wi-Fi RSSI without the location permission
            //                      and the entitlement on iOS 14+).
            //   is_expensive     — Wi-Fi Assist detection (Wi-Fi yếu → OS đi cellular)
            //   is_constrained   — Low Data Mode (iOS 13+)
            "network_type":     net["type"] ?? "unknown",
            "cellular_subtype": net["cellular_subtype"] ?? "",
            "network_label":    net["label"] ?? "",
            "network_strength": net["strength"] ?? "",
            "is_expensive":     net["is_expensive"] ?? "false",
            "is_constrained":   net["is_constrained"] ?? "false",
            "is_debug":       isDebug() ? "true" : "false",
            "is_rooted":      isJailbroken() ? "true" : "false",              // jailbreak on iOS
            // device_imei: iOS doesn't expose IMEI to apps since iOS 9. We
            // fill the schema field with identifierForVendor (UUID stable
            // per app vendor + install). Empty when iOS hasn't assigned one
            // yet (rare — pre-launch state).
            "device_imei":    deviceId,                                       // schema key (IDFV)
            "device_id":      deviceId,                                       // legacy alias
            "sdk_version":    "1.0.0",
        ]
        return jsonObject(fields)
    }

    /// Stable per-vendor install id (resets if all this vendor's apps removed).
    /// Used as the device_imei value since iOS doesn't expose the real IMEI.
    private static func identifierForVendor() -> String {
        UIDevice.current.identifierForVendor?.uuidString ?? ""
    }

    /// Marketing name (vd "iPhone 15 Pro") from the hardware id (vd "iPhone16,1").
    /// Apple deliberately keeps the mapping table out of public API, so we ship
    /// the snapshot of the table; unknown ids fall through to the raw hardware
    /// id so anomaly detectors can still spot a new SKU.
    private static func marketingDeviceName() -> String {
        let id = hardwareModel()
        if let mapped = MARKETING_NAME[id] { return mapped }
        // Simulator returns "i386" / "arm64" / "x86_64" — surface that so the
        // operator can tell sim sessions from real-device sessions.
        if id == "i386" || id == "arm64" || id == "x86_64" {
            return "Simulator (\(id))"
        }
        return id
    }

    /// Snapshot of Apple's hardware id → marketing name table. Last refreshed
    /// 2026-06; keep this list pruned to current sell-through models + the
    /// long tail of iPhone 8+ / iPad Air 3+ (older models are out of support
    /// targets here). Add new ids each WWDC.
    private static let MARKETING_NAME: [String: String] = [
        // iPhone (most-likely to appear first)
        "iPhone17,1": "iPhone 16 Pro Max",
        "iPhone17,2": "iPhone 16 Pro",
        "iPhone17,3": "iPhone 16",
        "iPhone17,4": "iPhone 16 Plus",
        "iPhone16,1": "iPhone 15 Pro",
        "iPhone16,2": "iPhone 15 Pro Max",
        "iPhone15,4": "iPhone 15",
        "iPhone15,5": "iPhone 15 Plus",
        "iPhone15,2": "iPhone 14 Pro",
        "iPhone15,3": "iPhone 14 Pro Max",
        "iPhone14,7": "iPhone 14",
        "iPhone14,8": "iPhone 14 Plus",
        "iPhone14,2": "iPhone 13 Pro",
        "iPhone14,3": "iPhone 13 Pro Max",
        "iPhone14,4": "iPhone 13 mini",
        "iPhone14,5": "iPhone 13",
        "iPhone13,1": "iPhone 12 mini",
        "iPhone13,2": "iPhone 12",
        "iPhone13,3": "iPhone 12 Pro",
        "iPhone13,4": "iPhone 12 Pro Max",
        "iPhone12,1": "iPhone 11",
        "iPhone12,3": "iPhone 11 Pro",
        "iPhone12,5": "iPhone 11 Pro Max",
        "iPhone12,8": "iPhone SE (2nd gen)",
        "iPhone14,6": "iPhone SE (3rd gen)",
        "iPhone11,2": "iPhone XS",
        "iPhone11,4": "iPhone XS Max",
        "iPhone11,6": "iPhone XS Max",
        "iPhone11,8": "iPhone XR",
        "iPhone10,3": "iPhone X",
        "iPhone10,6": "iPhone X",
        "iPhone10,1": "iPhone 8",
        "iPhone10,4": "iPhone 8",
        "iPhone10,2": "iPhone 8 Plus",
        "iPhone10,5": "iPhone 8 Plus",
        // iPad — flagship lines
        "iPad16,3": "iPad Pro 11-inch (M4)",
        "iPad16,4": "iPad Pro 11-inch (M4)",
        "iPad16,5": "iPad Pro 13-inch (M4)",
        "iPad16,6": "iPad Pro 13-inch (M4)",
        "iPad14,8": "iPad Air 11-inch (M2)",
        "iPad14,9": "iPad Air 11-inch (M2)",
        "iPad14,10": "iPad Air 13-inch (M2)",
        "iPad14,11": "iPad Air 13-inch (M2)",
        "iPad14,3": "iPad Pro 11-inch (4th gen)",
        "iPad14,4": "iPad Pro 11-inch (4th gen)",
        "iPad14,5": "iPad Pro 12.9-inch (6th gen)",
        "iPad14,6": "iPad Pro 12.9-inch (6th gen)",
        "iPad14,1": "iPad mini (6th gen)",
        "iPad14,2": "iPad mini (6th gen)",
        "iPad13,18": "iPad (10th gen)",
        "iPad13,19": "iPad (10th gen)",
    ]

    private static func isDebug() -> Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }

    /// One-shot snapshot via NWPathMonitor. Returns a dict the caller can
    /// merge into the device info bag — we report the primary transport plus
    /// the two NWPath signals that matter for streaming attribution:
    ///   • is_expensive   — Wi-Fi Assist on iOS silently switches to cellular
    ///                      when the Wi-Fi link can't carry data; the status
    ///                      bar still shows Wi-Fi but `path.isExpensive` flips
    ///                      to true. That's the field to trust over `type`.
    ///   • is_constrained — Low Data Mode (iOS 13+); user-toggled throttle.
    /// We start the monitor on a dedicated queue, wait briefly for the first
    /// path update, then cancel — keeping init synchronous + lifecycle clean.
    private static func networkSnapshot() -> [String: String] {
        let monitor = NWPathMonitor()
        let q = DispatchQueue(label: "unitrack.netcheck")
        let sem = DispatchSemaphore(value: 0)
        var captured: NWPath?
        monitor.pathUpdateHandler = { path in
            captured = path
            sem.signal()
        }
        monitor.start(queue: q)
        // Bounded wait: NWPathMonitor delivers the current path almost
        // immediately, but cap at 500ms so a stuck network daemon can't hang
        // app init.
        _ = sem.wait(timeout: .now() + .milliseconds(500))
        monitor.cancel()

        guard let path = captured else {
            return ["type": "unknown"]
        }
        if path.status != .satisfied {
            return ["type": "none"]
        }
        var out: [String: String] = [
            "is_expensive":   path.isExpensive ? "true" : "false",
            "is_constrained": path.isConstrained ? "true" : "false",
        ]
        // Wired check first — even on iOS this can be true for tethered
        // Ethernet adapters (rare but real). VPN must beat the others because
        // a VPN tunnel hides what the underlying transport actually is.
        if path.usesInterfaceType(.wiredEthernet) {
            out["type"]     = "wired"
            out["label"]    = "ethernet"
        } else if path.usesInterfaceType(.wifi) {
            out["type"]     = "wifi"
            out["label"]    = "wifi"
            // Wi-Fi RSSI estimate via Reachability/CoreFoundation isn't
            // possible from a sandboxed iOS app without the
            // com.apple.developer.networking.wifi-info entitlement (Apple
            // grants only to enterprise apps). We still try `wifiRSSI()` so
            // the field is filled when the entitlement is present.
            if let rssi = wifiRSSI() { out["strength"] = String(rssi) }
        } else if path.usesInterfaceType(.cellular) {
            let gen = cellularGeneration()
            out["type"]             = "cellular"
            out["cellular_subtype"] = gen
            // Friendly cellular label: "5G" / "4G" / "3G" / "2G". Capital so it
            // matches the wording carriers + UX teams use ("connected via 5G").
            out["label"] = gen.uppercased()
            if let bars = cellularSignalBars() { out["strength"] = String(bars) }
        } else if path.usesInterfaceType(.other) {
            // .other covers VPN tunnels + relayed traffic. Report it
            // explicitly so the downstream pipeline doesn't misread it as
            // wifi.
            out["type"]  = "vpn"
            out["label"] = "vpn"
        } else {
            out["type"]  = "unknown"
            out["label"] = "unknown"
        }
        return out
    }

    /// Wi-Fi RSSI in dBm (vd -55). Only works when the app has the
    /// `com.apple.developer.networking.wifi-info` entitlement; returns nil
    /// otherwise. We swallow the failure quietly — the field is documented as
    /// best-effort.
    private static func wifiRSSI() -> Int? {
        // CNCopyCurrentNetworkInfo deprecated since iOS 14; the modern path
        // is NEHotspotNetwork.fetchCurrent(...) but it's async + requires
        // the same entitlement. Returning nil here keeps the snapshot
        // synchronous; apps with the entitlement can supply RSSI via
        // UniTrack.setApplicationContext(...) at runtime.
        return nil
    }

    /// Cellular signal bars 0..4 from CoreTelephony. iOS doesn't expose RSSI
    /// in dBm to apps (private API), so bars is the closest stable proxy.
    /// nil when no carrier is reachable.
    private static func cellularSignalBars() -> Int? {
        // CoreTelephony exposes carrier metadata but not signal strength to
        // third-party apps. Same story as wifiRSSI: surface nil and let the
        // app override via setApplicationContext when it has access (vd
        // status-bar overlay private API in enterprise builds).
        return nil
    }

    /// Map the current radio access technology to a generation label.
    /// Falls back to "cellular" if the RAT can't be determined.
    private static func cellularGeneration() -> String {
        let info = CTTelephonyNetworkInfo()
        var rat: String?
        if #available(iOS 12.0, *) {
            // Pick the active carrier's RAT (first available entry).
            rat = info.serviceCurrentRadioAccessTechnology?.values.first
        } else {
            rat = info.currentRadioAccessTechnology
        }
        guard let tech = rat else { return "cellular" }
        switch tech {
        case CTRadioAccessTechnologyGPRS,
             CTRadioAccessTechnologyEdge,
             CTRadioAccessTechnologyCDMA1x:
            return "2g"
        case CTRadioAccessTechnologyWCDMA,
             CTRadioAccessTechnologyHSDPA,
             CTRadioAccessTechnologyHSUPA,
             CTRadioAccessTechnologyeHRPD,
             CTRadioAccessTechnologyCDMAEVDORev0,
             CTRadioAccessTechnologyCDMAEVDORevA,
             CTRadioAccessTechnologyCDMAEVDORevB:
            return "3g"
        case CTRadioAccessTechnologyLTE:
            return "4g"
        default:
            // 5G NR constants exist only on iOS 14.1+; compare by raw value so
            // this compiles on older SDKs too.
            if tech == "CTRadioAccessTechnologyNRNSA" || tech == "CTRadioAccessTechnologyNR" {
                return "5g"
            }
            return "cellular"
        }
    }

    /// Best-effort jailbreak detection (common file/sandbox checks).
    private static func isJailbroken() -> Bool {
        #if targetEnvironment(simulator)
        return false
        #else
        let paths = ["/Applications/Cydia.app", "/bin/bash", "/usr/sbin/sshd",
                     "/etc/apt", "/private/var/lib/apt/"]
        for p in paths where FileManager.default.fileExists(atPath: p) { return true }
        // Can we write outside the sandbox?
        let probe = "/private/jailbreak_probe.txt"
        if (try? "x".write(toFile: probe, atomically: true, encoding: .utf8)) != nil {
            try? FileManager.default.removeItem(atPath: probe)
            return true
        }
        return false
        #endif
    }

    /// e.g. "iPhone15,2" — the hardware identifier (not the marketing name).
    private static func hardwareModel() -> String {
        var sysinfo = utsname()
        uname(&sysinfo)
        let machine = withUnsafeBytes(of: &sysinfo.machine) { raw -> String in
            let ptr = raw.baseAddress!.assumingMemoryBound(to: CChar.self)
            return String(cString: ptr)
        }
        return machine
    }

    private static func str(_ v: Any?) -> String {
        (v as? String) ?? ""
    }

    private static func jsonObject(_ dict: [String: String]) -> String {
        // All values are strings; build manually to avoid pulling Codable here.
        if let data = try? JSONSerialization.data(
                withJSONObject: dict, options: [.sortedKeys]),
           let s = String(data: data, encoding: .utf8) {
            return s
        }
        return "{}"
    }
}
